File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    // BFS recursive
    // public int maxDepth(TreeNode root) {
    //     if (root == null) return 0;
    //     return (1 + Math.max(maxDepth(root.left), maxDepth(root.right)));
    // }

    // BFS iterative
    // public int maxDepth(TreeNode root) {
    //     if (root == null) return 0;
    //     Queue<TreeNode> q = new LinkedList<>();
    //     q.add(root);
    //     int result = 0;
    //     while(q.size() > 0){
    //         int len = q.size();
    //         for(int i = 0; i<len;i++){
    //             TreeNode temp = q.peek();
    //             q.remove();
    //             if (temp.left != null){
    //                 q.add(temp.left);
    //             }
    //             if (temp.right  != null){
    //                 q.add(temp.right);
    //             }
    //         }
    //         result += 1;
    //     }
    //     return result;
    // }

    // DFS iterative
    public int maxDepth(TreeNode root) {
        Stack<Pair<TreeNode, Integer>> s = new Stack<>();
        Pair<TreeNode, Integer> head  = new Pair<TreeNode, Integer>(root, 1);
        s.push(head);
        int result = 0;
        while (s.size() > 0 ){
            Pair<TreeNode, Integer> temp = s.peek();
            s.pop();
            if(temp.getKey() != null){
                if(temp.getKey().left != null){
                    Pair<TreeNode, Integer> t = new Pair<TreeNode, Integer>(temp.getKey().left, 1 + temp.getValue());
                    s.push(t);
                }
                if(temp.getKey().right != null){
                    Pair<TreeNode, Integer> t = new Pair<TreeNode, Integer>(temp.getKey().right, 1 + temp.getValue());
                    s.push(t);
                }          
                result = Math.max(result, temp.getValue());
            }
        }
        return result;
    }
}

File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include<queue>
#include<stack>
class Solution {
public:
    // BFS recursive
    // int maxDepth(TreeNode* root) {
    //     if (!root) return 0;
    //     return (1 + max(maxDepth(root->left), maxDepth(root->right)));
    // }

    // BFS iterative
    // int maxDepth(TreeNode* root) {
    //     if (!root) return 0;
    //     queue<TreeNode*> q;
    //     q.push(root);
    //     int result = 0;
    //     while(q.size()){
    //         cout << q.size() <<": ";
    //         int len = q.size();
    //         for(int i = 0; i<len;i++){
    //             TreeNode* temp = q.front();
    //             cout << temp->val << ", " ;
    //             q.pop();
    //             if (temp->left){
    //                 q.push(temp->left);
    //             }
    //             if (temp->right){
    //                 q.push(temp->right);
    //             }
    //         }
    //         cout<<"\n";
    //         result += 1;
    //     }
    //     return result;
    // }

    // DFS iterative
    int maxDepth(TreeNode* root) {
        stack<pair<TreeNode*, int>> s;
        pair<TreeNode*, int> head (root, 1);
        s.push(head);
        int result = 0;
        while (s.size()){
            pair<TreeNode*, int> temp = s.top();
            s.pop();
            if(temp.first){
                if(temp.first->left){
                    pair<TreeNode*, int> t (temp.first->left, 1 + temp.second);
                    s.push(t);
                }
                if(temp.first->right){
                    pair<TreeNode*, int> t (temp.first->right, 1 + temp.second);
                    s.push(t);
                }          
                result = max(result, temp.second);
            }
        }
        return result;
    }
};